### PR TITLE
Added details to docs for PR 2556

### PR DESCRIPTION
**Pull request by pdirt**
## Proposed changes

[//]: # Updating testrunner guide retry section

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [ x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ x] I have added necessary documentation (if appropriate)

## Further comments

[//]: #

### CODE REVIEW

**Comment by christian-bromann** [comment]
@pdirt could you also document this option in the following places:
https://github.com/webdriverio/webdriverio/blob/master/examples/wdio.conf.js
https://github.com/webdriverio/webdriverio/blob/master/docs/guide/getstarted/configuration.md
https://github.com/webdriverio/webdriverio/blob/master/lib/helpers/wdio.conf.ejs

**Comment by pdirt** [comment]
@christian-bromann no worries ... I need to run some errands but will get to it soon.

File: docs/guide/getstarted/configuration.md [hunk 1]
```diff
@@ -293,3 +293,6 @@ Similarly for mocha:
         requires: ['./test/helpers/common.js']
     },
 ```
+##specFileRetries
```
Comment (christian-bromann) [inline docs/guide/getstarted/configuration.md]: This should be moved up to the other options like `connectionRetryTimeout` and should have a similar format than these

File: docs/guide/testrunner/retry.md [hunk 2]
```diff
@@ -80,3 +80,28 @@ module.exports = function () {
 ```
 
 Reruns can only be defined in your step definitions file and not in your feature file. To use this you have to have the [wdio-cucumber-framework](https://github.com/webdriverio/wdio-cucumber-framework) adapter installed with `v0.1.0` or greater.
+
+## Add retries on a per-specfile basis
+
+Previously only test-level and suite-level retries were available, which
```
Comment (christian-bromann) [inline docs/guide/testrunner/retry.md]: Please no line breaks

File: docs/guide/getstarted/configuration.md (hunk 1 above)
Comment (pdirt) [reply docs/guide/getstarted/configuration.md]: Thanks, I did sign the CLA as far as I can tell.  I will fix these other issues asap.
https://cla.js.foundation/webdriverio/webdriverio?pullRequest=2614
->
"You have signed the CLA for webdriverio/webdriverio"

File: docs/guide/getstarted/configuration.md [hunk 3]
```diff
@@ -106,6 +106,9 @@ Default: *90000*
 ### connectionRetryCount
 Count of request retries to the Selenium server
 
+###specFileRetries
```
Comment (christian-bromann) [inline docs/guide/getstarted/configuration.md]: I don't think the type and default description below applies to the specFileRetries option also let's be consistent and keep a space between `###` and `specFileRetries`

**Comment by pdirt** [comment]
On the line breaks:

are you saying line 83 and 85? Because I was following what was done previously like on lines on 66/68. My apologies if I am missing something obvious. @christian-bromann

File: docs/guide/testrunner/retry.md (hunk 2 above)
Comment (pdirt) [reply docs/guide/testrunner/retry.md]: Doh, ok I see. fixing.

File: docs/guide/getstarted/configuration.md (hunk 3 above)
Comment (pdirt) [reply docs/guide/getstarted/configuration.md]: I think you may right I pinged @bennieswart because I am a little confused myself on this.

**Comment by bennieswart** [comment]
@pdirt perhaps you can PR this for `bennieswart:retries_per_specfile`, then we can get this merged into `webdriverio` in one PR. Definitely not critical, but seems neater.
@christian-bromann what is your preference on this?

File: docs/guide/testrunner/retry.md [hunk 4]
```diff
@@ -80,3 +80,20 @@ module.exports = function () {
 ```
 
 Reruns can only be defined in your step definitions file and not in your feature file. To use this you have to have the [wdio-cucumber-framework](https://github.com/webdriverio/wdio-cucumber-framework) adapter installed with `v0.1.0` or greater.
+
+## Add retries on a per-specfile basis
+
+Previously only test-level and suite-level retries were available, which are fine in most cases. In any test which involves state, such as on a server or in a db, the state may be left invalid once the test fails the first time. Any subsequent retries may have no chance of passing due to the invalid state they would have to start with.
+
+A new browser instance is created for each specfile, which makes this an ideal place to hook and setup any other states (server, db). Retries on this level would mean that the whole setup process could simply be repeated as if it were for a new specfile, with the results of the failing specfile (that is to be retried) being discarded.
```
Comment (bennieswart) [inline docs/guide/testrunner/retry.md]: Seeing as this has advanced from a suggestion to a feature, I'd say change `would mean that` to `means that`.

File: lib/helpers/wdio.conf.ejs [hunk 5]
```diff
@@ -134,6 +134,9 @@ exports.config = {
     //     webdriverrtc: {},
     //     browserevent: {}
     // },
+    // Set retry tests on a spec file basis 
+    // Each failed test in a suite gets a new environment before the test is retried
+    specFileRetries:
```
Comment (bennieswart) [inline lib/helpers/wdio.conf.ejs]: Missing ` 1,` at end of line.

File: docs/guide/getstarted/configuration.md [hunk 6]
```diff
@@ -106,8 +106,11 @@ Default: *90000*
 ### connectionRetryCount
 Count of request retries to the Selenium server
 
+### specFileRetries
+Sets the testrunner to retry a failed spec in a suite x amount of times, the retried spec gets a new setup environment for its new test.
+
 Type: `Number`<br>
-Default: *3*
+Default: *0*
```
Comment (christian-bromann) [inline docs/guide/getstarted/configuration.md]: Can we add back

Comment (christian-bromann) [inline docs/guide/getstarted/configuration.md]: Can we add back
```md
Type: `Number`<br>
Default: *3*
```

Comment (pdirt) [reply docs/guide/getstarted/configuration.md]: no worries, fixed.

Comment (christian-bromann) [reply docs/guide/getstarted/configuration.md]: @pdirt did you push it? Still see old state

**Comment by pdirt** [comment]
https://github.com/webdriverio/webdriverio/pull/2614/commits/a932fcdd7e3803889761cabd33373c7e4a3f89c4

Should be there now.

-p

**Comment by pmerwin** [comment]
Awesome!!!  Thanks!

-p

**Comment by christian-bromann** [comment]
This PR needs to be rebased on top of master and the CLA needs to get signed by @pdirt

**Comment by pmerwin** [comment]
Hmm not sure why it thinks i have not signed the CLA, see the attachment...
it says I have signed it, but if you . look at the PR it does not agree?

-p

On Tue, May 15, 2018 at 3:58 AM, Christian Bromann <notifications@github.com

**Comment by christian-bromann** [comment]
> look at the PR it does not agree?

You need to sign it with the same email that you've used to commit

**Comment by pmerwin** [comment]
I see that, but when I try to change it it won't let me?   When I hit the
CLA link it shows I signed it with my other account, but am logged in with
the one that I want to sign it with, but I see no option to do that.

On Tue, May 15, 2018 at 8:29 AM, Christian Bromann <notifications@github.com

**Comment by christian-bromann** [comment]
@pmerwin try with an anonymous browser session

**Comment by pmerwin** [comment]
@christian-bromann I rebased per:

https://gist.github.com/ravibhure/a7e0918ff4937c9ea1c456698dcd58aa

did I do it right?

**Pull request closed without merging**